numpy/core/numeric.py

```diff
@@ -373,13 +377,113 @@ def extend_all(module):
+    (recursively) counts how many elements in ``a`` (and in
+    sub-arrays thereof) have their '__nonzero__()' or '__bool__()'
+    method evaluated to 'True'.
+
+    Parameters
+    ----------
+    a : array_like
+        The array for which to count non-zeros.
+
+    axis : int or tuple, optional
+        Axis or tuple of axes along which to count non-zeros.
+        Default is None, meaning that non-zeros will be counted
+        along a flattened version of ``a``.
+
+        .. versionadded:: 1.12.0
+
+    Returns
+    -------
+    count : int or array of int
+        Number of non-zero values in the array along a given axis.
+        Otherwise, the total number of non-zero values in the array
+        is returned.
+
+    See Also
+    --------
+    nonzero : Return the coordinates of all the non-zero values.
+
+    Examples
+    --------
+    >>> np.count_nonzero(np.eye(4))
+    4
+    >>> np.count_nonzero([[0,1,7,0,0],[3,0,0,2,19]])
+    5
+    >>> np.count_nonzero([[0,1,7,0,0],[3,0,0,2,19]], axis=0)
+    array([1, 1, 1, 1, 1])
+    >>> np.count_nonzero([[0,1,7,0,0],[3,0,0,2,19]], axis=1)
+    array([2, 3])
+
+    """
+    from numpy import (apply_along_axis, atleast_1d,
+                       intp, issubdtype, string_,
+                       number, unicode_, unique)
+
+    a = asanyarray(a)
+
+    # The following two checks ensure consistent behaviour
+    # within count_nonzero and between other numpy functions
+    # and ufuncs that accept axis parameters
+    if hasattr(axis, '__len__') and type(axis) != tuple:
+        raise TypeError("A tuple or integer is required")
+
+    if not axis and axis != 0:
+        axis = None
+
+    # multiarray is faster with
+    # booleans when axis is None
+    if a.dtype == bool:
+        return (multiarray.count_nonzero(a) if axis is None
+                else a.sum(axis=axis, dtype=intp))
+
+    # sum is faster with numbers
+    # independent of the axis
+    if issubdtype(a.dtype, number):
+        return (a != 0).sum(axis=axis, dtype=intp)
+
+    # multiarray is faster with
+    # strings when axis is None
+    if issubdtype(a.dtype, string_) or issubdtype(a.dtype, unicode_):
+        nullstr = a.dtype.type('')
+        return (multiarray.count_nonzero(a) if axis is None
+                else (a != nullstr).sum(axis=axis, dtype=intp))
+
+    if axis is None:
+        return multiarray.count_nonzero(a)
+
+    axis = atleast_1d(axis)
+
+    if ((axis.ndim > 1 and axis.shape[0] != 1) or
+            not issubdtype(axis.dtype, number)):
+        raise ValueError(('Invalid axis provided. Cannot '
+                          'convert to scalar index.'), axis)
+
+    if unique(axis).size != axis.size:
+        raise ValueError("Duplicate value in 'axis'")
+
+    axis = axis.flatten()
+    counts = apply_along_axis(multiarray.count_nonzero, axis[0], a)
+
+    return (counts if axis.size == 1 else
+            counts.sum(axis=tuple(axis[1:] - (
+                axis[1:] > axis[0])), dtype=intp))
```

Comment:
This expression should absolutely be written out in several lines for clarity. Also, please explain the trick you're using for the axis argument in a comment -- it's not obvious to me what is going on there.